tsl/src/compression/compression.c

```diff
@@ -2184,57 +2259,47 @@ fill_predicate_context(Chunk *ch, List *predicates, List **segmentby_columns,
  * OUT param null_columns is saved with column attribute number.
  */
 static ScanKeyData *
-build_update_delete_scankeys(Chunk *chunk, List *segmentby_columns, List *segmentby_column_values,
-							 List *is_null_check, int *num_scankeys, Bitmapset **null_columns)
+build_update_delete_scankeys(RowDecompressor *decompressor, List *filter_columns, List *filter_ops,
+							 List *segmentby_column_values, List *is_null_check, int *num_scankeys,
+							 Bitmapset **null_columns)
 {
-	Chunk *comp_chunk = NULL;
-	Relation comp_chunk_rel;
-	TupleDesc tupleDesc;
 	ListCell *col;
 	ListCell *is_null_or_not_null;
 	int key_index = 0;
 	int pos = 0;
 
-	ScanKeyData *scankeys = palloc0(segmentby_columns->length * sizeof(ScanKeyData));
-	comp_chunk = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
-	comp_chunk_rel = table_open(comp_chunk->table_id, AccessShareLock);
-	tupleDesc = RelationGetDescr(comp_chunk_rel);
+	ScanKeyData *scankeys = palloc0(filter_columns->length * sizeof(ScanKeyData));
 
-	forboth (col, segmentby_columns, is_null_or_not_null, is_null_check)
+	forboth (col, filter_columns, is_null_or_not_null, is_null_check)
 	{
 		char *column_name = lfirst(col);
-		AttrNumber attnum = get_attnum(RelationGetRelid(comp_chunk_rel), column_name);
-		if (attnum == InvalidAttrNumber)
+		AttrNumber attno = get_attnum(decompressor->in_rel->rd_id, column_name);
+		if (attno == InvalidAttrNumber)
 			ereport(ERROR,
 					(errcode(ERRCODE_UNDEFINED_COLUMN),
 					 errmsg("column \"%s\" of relation \"%s\" does not exist",
 							column_name,
-							RelationGetRelationName(comp_chunk_rel))));
-		Form_pg_attribute attr = TupleDescAttr(tupleDesc, AttrNumberGetAttrOffset(attnum));
-		TypeCacheEntry *tce = lookup_type_cache(attr->atttypid, TYPECACHE_EQ_OPR_FINFO);
-		if (!OidIsValid(tce->eq_opr_finfo.fn_oid))
-			elog(ERROR, "no equality function for column \"%s\"", column_name);
-
-		if (lfirst_int(is_null_or_not_null) == 0)
+							RelationGetRelationName(decompressor->in_rel))));
+		bool isnull = lfirst_int(is_null_or_not_null) == 1;
+		Datum value = 0;
+		StrategyNumber strategy = 0;
+		if (!isnull)
 		{
-			Const *const_value = list_nth(segmentby_column_values, pos++);
-			Datum value = (const_value ? const_value->constvalue : 0);
-			ScanKeyEntryInitialize(&scankeys[key_index++],
-								   0, /* flags */
-								   attnum,
-								   BTEqualStrategyNumber,
-								   InvalidOid, /* No strategy subtype. */
-								   attr->attcollation,
-								   tce->eq_opr_finfo.fn_oid,
-								   value);
-		}
-		else
-		{
-			*null_columns = bms_add_member(*null_columns, attnum);
+			Const *const_value = list_nth(segmentby_column_values, pos);
+			value = (const_value ? const_value->constvalue : 0);
+			strategy = list_nth_int(filter_ops, pos++);
 		}
+
+		key_index = create_segment_metadata_scankey(decompressor,
```

Comment:
This function not only creates scankeys for compresssion metadata columns like _ts_meta_min/max_1, it is also used to create scankeys for segment by columns (which is not metadata column).
My be rename the function to be more generic.